File: 2022/days/13/main.py
from typing import List, Tuple, TypeVar, Union
from functools import cmp_to_key

T = TypeVar('T')
NestedList = Union[T, List['NestedList[T]']]


def extractNestedList(line: str) -> NestedList[int]:
    nestedList: NestedList[int] = []

    toadd: str = ''

    index = 0
    while (index + 1 <= len(line)):
        character = line[index]

        if not character.isnumeric() and toadd != '':
            c = int(toadd)
            nestedList.append(c)
            toadd = ''

        if character == ',':
            index = index + 1
            continue

        if character == '[':
            endIndex = chopchop(line, index)
            subline = line[index+1:endIndex-1]
            subList = extractNestedList(subline)
            nestedList.append(subList)
            index = endIndex
            continue

        if character == ']':
            index = index + 1
            continue

        if character.isnumeric():
            toadd = toadd + character
            index = index + 1
            continue

    if toadd != '' and toadd.isnumeric():
        nestedList.append(int(toadd))

    return nestedList


def chopchop(line: str, startingIndex: int = 0, leftDelim: str = '[', rightDelim: str = ']'):
    """Returns the index that our substring has the same number of leftDelims and rightDelims.  

    IE, returns the index that we have the same number of left and right parenthesies"""
    depth = 0
    for index, character in enumerate(line[startingIndex:]):
        if character == leftDelim:
            depth = depth + 1

        if character == rightDelim:
            depth = depth - 1
            if depth == 0:
                return startingIndex + index + 1

    return -1


def readFile(file) -> List[NestedList[int]]:
    nestedLists: List[NestedList[int]] = []

    for line in file:
        line = line.strip('\n')
        if line == '':
            continue
        line = line[1:]
        list = extractNestedList(line)

        nestedLists.append(list)

    return nestedLists


def orderNestedLists(left, right):
    def compare(l, r):
        if l < r:
            return True
        if l > r:
            return False
        return None

    if type(left) == int and type(right) == int:
        return compare(left, right)

    if type(left) == int:
        left = [left]
    if type(right) == int:
        right = [right]

    for index in range(min(len(left), len(right))):
        result = orderNestedLists(left[index], right[index])
        if result is not None:
            return result

    return compare(len(left), len(right))


def partOne(lists: List[NestedList]) -> int:
    pairedLists = zip(lists[::2], lists[1::2])
    acc = 0

    for index, (left, right) in enumerate(pairedLists, start=1):
        ordered = orderNestedLists(left, right)
        if ordered:
            acc += index

    return acc


def partTwo(lists: List[NestedList]) -> int:

    def compareFuction(l, r):
        result = orderNestedLists(l, r)
        if result is True:
            return 1
        if result is False:
            return -1
        return 0

    divider_packets = [[[2]], [[6]]]

    for packet in divider_packets:
        lists.append(packet)

    lists.sort(key=cmp_to_key(compareFuction), reverse=True)

    acc: int = 1
    for packet in divider_packets:
        i = lists.index(packet) +1
        acc *= i
    return acc


if __name__ == "__main__":
    # for debugging these should be specified as absolute paths, for running they can just be specified as relative
    filename = r'example.txt'
    filename = r'input.txt'


    with open(filename, 'r') as fp:
        lists = readFile(fp)

    print(partOne(lists))
    print(partTwo(lists))

File: 2022/days/13/main.ts
import { readFileSync } from 'fs'
type NestedArray<T> = Array<NestedArray<T>> | T;

function getSubclauseEndIndex(input: string, startingIndex: number, leftDelim: string = '[', rightDelim: string = ']'): number {
    let depth: number = 0

    for (let index = startingIndex; index < input.length; index++) {
        const element = input[index];

        if (element == leftDelim) {
            depth += 1
            continue
        }

        if (element == rightDelim) {
            depth -= 1

            if (depth == 0) return index + 1
        }

    }

    return -1
}

function extractNestedList(input: string, leftDelim: string = '[', rightDelim: string = ']'): NestedArray<number> {
    let nestedArray: NestedArray<number> = []
    let toAdd: string = ''

    for (let index = 0; index < input.length; index++) {
        const character: string = input[index];

        if (!isNaN(parseInt(character))) {
            toAdd += character
        } else if (toAdd != '') {
            nestedArray.push(parseInt(toAdd))
            toAdd = ''
        }

        if (character == leftDelim) {
            const endIndex: number = getSubclauseEndIndex(input, index, leftDelim, rightDelim)
            const substring: string = input.slice(index + 1, endIndex - 1)
            nestedArray.push(extractNestedList(substring, leftDelim, rightDelim))
            input = input.slice(endIndex)
        }
    }

    if (toAdd != '') {
        nestedArray.push(parseInt(toAdd))
        toAdd = ''
    }

    return nestedArray
}

function loadLists(input: string): NestedArray<number>[] {
    const splitString: string[] = fs.split('\n')
    let lists: NestedArray<number>[] = []
    lists.pop()

    splitString.forEach(line => {
        line = line.replace("\r", '')
        if (line != '') lists.push(extractNestedList(line))
    })

    return lists
}



function ordered(left: NestedArray<number>, right: NestedArray<number>): number {
    function cmp(left: number, right: number): number {
        return right - left
    }

    if (typeof (left) == 'number' && typeof (right) == 'number') return cmp(left, right)

    if (typeof (left) == 'number') left = [left]
    if (typeof (right) == 'number') right = [right]

    const minlen = Math.min(left.length, right.length)

    for (let index = 0; index < minlen; index++) {
        const res = ordered(left[index], right[index])
        if (res != 0) return res
    }

    return cmp(left.length, right.length)
}

function indexOf(lists: NestedArray<number>[], value: NestedArray<number>): number {
    for (let index = 0; index < lists.length; index++) {
        const element = lists[index];
        if (element == value) return index
    }
    return -1
}

function partOne(lists: NestedArray<number>[]): number {
    // my kingdom for list[::2] and list[1::2]
    const firsts = lists.filter((element, index) => index % 2 == 0)
    const seconds = lists.filter((element, index) => (index + 1) % 2 == 0)
    let pairedLists: NestedArray<number>[][] = []
    pairedLists.pop()

    for (let i = 0; i < firsts.length; i++) {
        const f = firsts[i];
        const s = seconds[i]
        pairedLists.push([f, s])
    }

    let acc = 0
    pairedLists.forEach((pair, index) => {
        const sorted = ordered(pair[0], pair[1])
        if (sorted != null && sorted > 0) acc += index + 1
    });

    return acc

}

function partTwo(lists: NestedArray<number>[], dividerPackets: NestedArray<number>[] = [[[2]],[[6]]]): number {

    dividerPackets.forEach(packet => {
        lists.push(packet)
    });

    lists.sort(ordered).reverse()

    let acc = 1
    dividerPackets.forEach(packet => {
        acc *= indexOf(lists,packet) +1
    });

    return acc

}


let fs = readFileSync('input.txt', 'utf-8');
let lists: NestedArray<number>[] = loadLists(fs);

console.log(partOne(lists))
console.log(partTwo(lists))
